src/main/host/descriptor/eventfd.rs

```diff
@@ -0,0 +1,224 @@
+    FileMode, FileState, FileStatus, StateEventSource, StateListenerFilter,
+};
+use crate::host::memory_manager::MemoryManager;
+use crate::host::syscall_types::{PluginPtr, SyscallResult};
+use crate::utility::event_queue::{EventQueue, Handle};
+use crate::utility::stream_len::StreamLen;
+
+pub struct EventFdFile {
+    counter: u64,
+    semaphore_mode: bool,
+    event_source: StateEventSource,
+    state: FileState,
+    status: FileStatus,
+}
+
+impl EventFdFile {
+    pub fn new(init_value: u64, semaphore_mode: bool, status: FileStatus) -> Self {
+        Self {
+            counter: init_value,
+            semaphore_mode,
+            event_source: StateEventSource::new(),
+            state: FileState::ACTIVE | FileState::WRITABLE,
+            status,
+        }
+    }
+
+    pub fn get_status(&self) -> FileStatus {
+        self.status
+    }
+
+    pub fn set_status(&mut self, status: FileStatus) {
+        self.status = status;
+    }
+
+    pub fn mode(&self) -> FileMode {
+        FileMode::READ | FileMode::WRITE
+    }
+
+    pub fn close(&mut self, event_queue: &mut EventQueue) -> SyscallResult {
+        // set the closed flag and remove the active, readable, and writable flags
+        self.copy_state(
+            FileState::CLOSED | FileState::ACTIVE | FileState::READABLE | FileState::WRITABLE,
+            FileState::CLOSED,
+            event_queue,
+        );
+
+        Ok(0.into())
+    }
+
+    pub fn read<W>(
+        &mut self,
+        mut bytes: W,
+        offset: libc::off_t,
+        event_queue: &mut EventQueue,
+    ) -> SyscallResult
+    where
+        W: std::io::Write + std::io::Seek,
+    {
+        // eventfds don't support seeking
+        if offset != 0 {
+            return Err(Errno::ESPIPE.into());
+        }
+
+        // eventfd(2): "Each successful read(2) returns an 8-byte integer"
+        const NUM_BYTES: usize = 8;
+
+        // this check doesn't guarentee that we can write all bytes since the stream length is only
+        // a hint
+        if usize::try_from(bytes.stream_len_bp()?).unwrap() < NUM_BYTES {
+            log::trace!(
+                "Reading from eventfd requires a buffer of at least {} bytes",
+                NUM_BYTES
+            );
+            return Err(Errno::EINVAL.into());
+        }
+
+        if self.counter == 0 {
+            log::trace!("Eventfd counter is 0 and cannot be read right now");
+            return Err(Errno::EWOULDBLOCK.into());
+        }
+
+        // behavior defined in `man 2 eventfd`
+        if self.semaphore_mode {
+            const ONE: [u8; NUM_BYTES] = 1u64.to_ne_bytes();
+            bytes.write_all(&ONE)?;
+            self.counter -= 1;
+        } else {
+            let to_write: [u8; NUM_BYTES] = self.counter.to_ne_bytes();
+            bytes.write_all(&to_write)?;
+            self.counter = 0;
+        }
+
+        self.update_state(event_queue);
+
+        return Ok(NUM_BYTES.into());
+    }
+
+    pub fn write<R>(
```

Comment:
Nit: Does rust let us use more than one letter for the type name? Would something like `ReadSeek` or `RdSk` be more descriptive?